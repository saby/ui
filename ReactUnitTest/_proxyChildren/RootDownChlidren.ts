import { Control, IControlOptions, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_proxyChildren/RootDownChlidren';

interface IRootDownChlidrenOptions extends IControlOptions {
    resolveAfterMount: (instance: RootDownChlidren) => void;
}

export default class RootDownChlidren extends Control<IRootDownChlidrenOptions> {
    protected _template: TemplateFunction = template;
    protected _value: number = 1;
    upValue() {
        this._value++;
    }
    _afterMount(): void {
        this._options.resolveAfterMount(this);
    }
}
