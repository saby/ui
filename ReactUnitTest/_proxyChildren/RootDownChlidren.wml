<div class="rootDownChlidren" name="testChildren">
   <ReactUnitTest._proxyChildren.DownControl
      children="{{ _children }}"
      value="{{ _value }}"
   />
</div>
