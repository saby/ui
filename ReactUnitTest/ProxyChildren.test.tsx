/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { screen } from '@testing-library/dom';
import RootDownChlidren from './_proxyChildren/RootDownChlidren';

describe('ProxyChildren', () => {
    let container: HTMLElement;
    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });
    afterEach(() => {
        jest.restoreAllMocks();
        unmountComponentAtNode(container);
        container.remove();
    });

    test('Нет спама варнингами, если _children прокидывают в дочерний шаблон или контрол', async () => {
        const warnSpy = jest.spyOn(console, 'warn');
        let resolveAfterMount: (instance: RootDownChlidren) => void;
        const promise = new Promise<RootDownChlidren>((resolve) => {
            resolveAfterMount = resolve;
        });
        act(() => {
            render(<RootDownChlidren resolveAfterMount={resolveAfterMount} />, container);
        });

        const instance = await promise;
        act(() => {
            instance.upValue();
        });
        await screen.findByText('value is 2');

        expect(container).toMatchSnapshot();
        expect(warnSpy).not.toHaveBeenCalled();
    });
});
