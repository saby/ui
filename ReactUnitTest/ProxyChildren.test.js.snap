// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProxyChildren Нет спама варнингами, если _children прокидывают в дочерний шаблон или контрол 1`] = `
<div>
  <div
    class="rootDownChlidren"
    name="testChildren"
    tabindex="0"
  >
    <div
      class="downControl"
      tabindex="0"
    >
      <div
        class="downTemplate"
        tabindex="0"
      >
        <span
          class="value"
        >
          value is 2
        </span>
        <span
          class="children"
        >
          rootDownChlidren
        </span>
      </div>
    </div>
  </div>
</div>
`;
